lib/full-site-editing/template-loader.php

```diff
@@ -64,73 +64,47 @@ function get_template_hierarchy( $template_type ) {
-				$block_template     = _gutenberg_get_template_file( 'wp_template', $current_block_template_slug );
-				$has_block_template = false;
-
-				if ( null !== $block_template && wp_get_theme()->get_stylesheet() === $block_template['theme'] ) {
-					$has_block_template = true;
-				}
-				// and that a corresponding block template from the theme and not the parent doesn't exist.
-				if ( $has_php_template && ! $has_block_template ) {
-					return $template;
-				}
-			}
-
-			break;
-		}
-
-		// Is this a custom template?
-		// This check should be removed when merged in core.
-		// Instead, wp_templates should be considered valid in locate_template.
-		$is_custom_template = 0 === strpos( $current_block_template_slug, 'wp-custom-template-' );
-
-		// Don't override the template if we find a template matching the slug we look for
-		// and which does not match a block template slug.
-		if (
-			! $is_custom_template &&
-			$current_template_slug !== $current_block_template_slug &&
-			$current_template_slug === $template_item_slug
-		) {
-			return $template;
-		}
+	if ( $template ) {
+		// locate_template() has found a PHP template at the path specified by $template.
+		// That means that we have a fallback candidate if we cannot find a block template
+		// with higher specificity.
+		// Thus, before looking for matching block themes, we shorten our list of candidate
+		// templates accordingly.
+
+		// Locate the index of $template (without the theme directory path) in $templates.
+		$relative_template_path = str_replace(
+			array( get_stylesheet_directory() . '/', get_template_directory() . '/' ),
+			'',
+			$template
+		);
+		$index                  = array_search( $relative_template_path, $templates, true );
+
+		// If the template hiearchy algorithm has successfully located a PHP template file,
+		// we will only consider block templates with higher or equal specificity.
+		$templates = array_slice( $templates, 0, $index + 1 );
 	}
 
-	if ( $current_template ) {
-		if ( empty( $current_template->content ) && is_user_logged_in() ) {
+	$block_template = gutenberg_resolve_template( $type, $templates );
+
+	if ( $block_template ) {
+		if ( empty( $block_template->content ) && is_user_logged_in() ) {
 			$_wp_current_template_content =
 			sprintf(
 				/* translators: %s: Template title */
 				__( 'Empty template: %s', 'gutenberg' ),
-				$current_template->title
+				$block_template->title
 			);
-		} elseif ( ! empty( $current_template->content ) ) {
-			$_wp_current_template_content = $current_template->content;
+		} elseif ( ! empty( $block_template->content ) ) {
+			$_wp_current_template_content = $block_template->content;
 		}
 		if ( isset( $_GET['_wp-find-template'] ) ) {
-			wp_send_json_success( $current_template );
+			wp_send_json_success( $block_template );
 		}
 	} else {
-		if ( 'index' === $type ) {
+		if ( $template ) {
+			return $template;
+		} elseif ( 'index' === $type ) {
```

Comment:
nitpick: `elseif` is not needed here, it can be a simple `if` since in the previous line we `return`

Reply:
381cfccf7c